lib/runnable.js

```diff
@@ -83,7 +83,10 @@ Runnable.prototype.timeout = function(ms) {
 
   // see #1652 for reasoning
   if (ms === range[0] || ms === range[1]) {
-    this._enableTimeouts = false;
+    this.enableTimeouts(false);
+  } else {
+    // Reset now that a new timeout limit has been set
+    this.enableTimeouts(true);
```

Comment:
This is how I think it should work:

A test calls `this.timeout(100)`.  The timeout should be set to 100ms.
1. A test subsequently calls `this.timeout(0)`, the timeout should _effectively_ be disabled.  Thereafter:
    - Calling `this.timeout(100)` again would re-enable the timeout
    - Calling `this.enableTimeouts(true)` would _either_ a) be a no-op, or b) should use the default timeout
1. A test subsequently calls `this.enableTimeouts(false)`, the timeout should be disabled.  Thereafter:
    - If a test calls `this.enableTimeouts(true)`, the timeout should be re-enabled and set to 100ms
    - If a test calls `this.timeout(200)`, this will set the internal timeout value to 200ms, _but timeouts will still be disabled; in order to re-enable timeouts, `this.enableTimeouts(true)` must _explicitly be called_.

I realize, currently, the values are conflated.  They should be separate, but `Runnable` should know that a `!this._timeout` _or_ `!_enableTimeouts` will cause timeouts to be ignored.

What do you think?

cc @juergba

Reply:
Fundamentally, I think calling `this.enableTimeouts(false)` then calling `this.timeout(500)` would violate the principal of least surprise.  But I could be wrong.

Reply:
Think I see what you mean, so value and state are 2 different things. 
- State: takes its priority from the last `enableTimeout()` only falling back to last `timeout` if none exist.
- Value: always comes from last `timeout()`.

Reply:
We should deprecate/remove all this `enableTimeouts()` and `_enableTimeouts` stuff.
And use the functionality as this user does. `this.skip(0)` for disabling and `this.skip(10)` for enabling. 
It's just a simple feature, there shouldn't be any potential surprises.

Reply:
(skip? assuming you mean `this.timeout()`). 

> It's just a simple feature, there shouldn't be any potential surprises. 

I'll have a go

Reply:
> We should deprecate/remove all this enableTimeouts() and _enableTimeouts stuff.

As we still want to keep the notion of "disabling" it seems easiest if we keep the private `_enableTimeouts` stuff but remove the public-facing side.

Reply:
Ok i've had a go, let me know what you guys think.

Reply:
if we're going to remove it, IMO it should be soft-deprecated for a major.

but, I think it's important to understand why it exists (see also: [Chesterton's Fence](https://en.wikipedia.org/wiki/Wikipedia:Chesterton's_fence)).  Mocha is full of these things.

from what I can tell, though, it doesn't look like it was a problem that couldn't have been solved via `--timeout 0` (see #1262).  IIRC, at the time, that didn't actually work properly, and you'd need to pass `--timeout <big-number>` instead.

an alternative which would avoid `semver-major` is:

1. leave `enableTimeouts(false)` in but make it an alias for `timeout(0)`
1. when `enableTimeouts(true)` is called, look at the parent `_timeout` until it finds a nonzero one; if it doesn't, choose the default (2000)
1. remove `_enableTimeouts`

Reply:
I can also see a valid use-case in disabling timeouts temporarily and then restoring to their last value.  you could have `--timeout 1000` on the command line, then someone wants to disable with `this.timeout(0)`, then later restore it, but do you also need to put `timeout(1000)` in your test file?  IMO the nicest thing to do here for the user would be to store the old value and restore it when calling `enableTimeouts(true)`.  so, I guess I kind of _like_ `enableTimeouts()`, but I don't like not having a single source of truth.

Reply:
I don't feel very strongly about this, though would prefer not breaking stuff, or at least providing an intend-to-break notice 😄

Reply:
I like clear cuts. And to keep simple things simple. I admit it's a sharp cut, because there hasn't been any deprecation message before removing, but it's ok. As @craigtaub mentioned `enableTimeouts()` is not a documented feature in our user docs.
We don't use `enableTimeouts()` internally neither, eg. we use `timeout(0)` when disabling timeout with Node option `inspect`.
I will review later.

Reply:
Thanks for having a look guys. To add to @juergba points:
- For the codebase it will be simpler (and easier) to maintain a single value for this throughout.
- For the users I can’t find a single mention of the “restore previous timeout value” feature anywhere in our docs (that includes API docs), just the overriding functionality. Due to that it seems unlikely to be well used (unless the user has gone through our own tests to learn more about it).

Given the above it appears as though a sharp cut should be fine, as it is good for us (roll out the change in 1 go) and good for our users (fixes the bug with overrides).
Thoughts?

Reply:
You could add a note in our [docs](https://mochajs.org/#timeouts):
As of v8.0.0 `enableTimeouts()` has been removed, ...

Reply:
The coverage decreased by your docs update. Does that mean our docs files are included in the coverage calculation? Which would be nonsense.

Reply:
1. `enableTimeouts()` [is documented](https://mochajs.org/api/mocha#enableTimeouts), but it is not mentioned in the "main" docs, fwiw.
1. @craigtaub pls rebase onto `master` and add `docs/**` to the `exclude` prop of `.nycrc`